pkg/driver/controller.go

```diff
@@ -0,0 +1,359 @@
+		}
+	}
+
+	if value, ok := volumeParams[GidMax]; ok {
+		// Ensure GID min is provided with GID max
+		if gidMin == 0 {
+			return nil, status.Errorf(codes.InvalidArgument, "Missing %v parameter", GidMin)
+		}
+		gidMax, err = strconv.Atoi(value)
+		if err != nil {
+			return nil, status.Errorf(codes.InvalidArgument, "Failed to parse invalid %v: %v", GidMax, err)
+		}
+		if gidMax <= gidMin {
+			return nil, status.Errorf(codes.InvalidArgument, "%v must be greater than %v", GidMax, GidMin)
+		}
+	} else {
+		// Ensure GID max is provided with GID min
+		if gidMin != 0 {
+			return nil, status.Errorf(codes.InvalidArgument, "Missing %v parameter", GidMin)
+		}
+	}
+
+	// Assign default GID ranges if not provided
+	if gidMin == 0 && gidMax == 0 {
+		gidMin = DefaultGidMin
+		gidMax = DefaultGidMax
+	}
+
+	if value, ok := volumeParams[DirectoryPerms]; ok {
+		accessPointsOptions.DirectoryPerms = value
+	}
+
+	if value, ok := volumeParams[BasePath]; ok {
+		basePath = value
+	}
+
+	// Check if file system exists. Describe FS handles appropriate error codes
+	if _, err = d.cloud.DescribeFileSystem(ctx, accessPointsOptions.FileSystemId); err != nil {
+		if err == cloud.ErrAccessDenied {
+			return nil, status.Errorf(codes.Unauthenticated, "Access Denied. Please ensure you have the right AWS permissions: %v", err)
+		}
+		if err == cloud.ErrNotFound {
+			return nil, status.Errorf(codes.InvalidArgument, "File System does not exist: %v", err)
+		}
+		return nil, status.Errorf(codes.Internal, "Failed to fetch File System info: %v", err)
+	}
+	gid, err := d.gidAllocator.getNextGid(accessPointsOptions.FileSystemId, gidMin, gidMax)
+	if err != nil {
+		return nil, err
+	}
+
+	gidStr := strconv.Itoa(gid)
+	rootDirName := RootDirPrefix + "-" + gidStr + "-" + uuid.New().String()
+	rootDir := basePath + "/" + rootDirName
+
+	accessPointsOptions.Gid = int64(gid)
+	accessPointsOptions.Uid = int64(gid)
+	accessPointsOptions.DirectoryPath = rootDir
+
+	accessPointId, err := d.cloud.CreateAccessPoint(ctx, volName, accessPointsOptions)
+	if err != nil {
+		d.gidAllocator.releaseGid(accessPointsOptions.FileSystemId, gid)
+		if err == cloud.ErrAccessDenied {
+			return nil, status.Errorf(codes.Unauthenticated, "Access Denied. Please ensure you have the right AWS permissions: %v", err)
+		}
+		if err == cloud.ErrAlreadyExists {
+			return nil, status.Errorf(codes.AlreadyExists, "Access Point already exists")
+		}
+		return nil, status.Errorf(codes.Internal, "Failed to create Access point in File System %v : %v", accessPointsOptions.FileSystemId, err)
+	}
+
+	return &csi.CreateVolumeResponse{
+		Volume: &csi.Volume{
+			CapacityBytes: volSize,
+			VolumeId:      accessPointsOptions.FileSystemId + "::" + accessPointId.AccessPointId,
+			VolumeContext: map[string]string{},
+		},
+	}, nil
+}
+
+func (d *Driver) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest) (*csi.DeleteVolumeResponse, error) {
+	klog.V(4).Infof("DeleteVolume: called with args %+v", *req)
```

Comment:
the gid is not released in this function

Reply:
This is by design, idea is to not re-allocate a GID once it is used

Reply:
but if the driver gets restarted incidentally ( rescheduled, or updated) then the GID will be reused anyway, so this has no real security benefit.

Also this introduces an artifical limit on # of volumes that can be provisioned over the driver's lifetime which needs to be documented. Of course it may take a long time to hit 7million volumes if the user keeps the gid parameters at the default but it is a possibility that must be documented

Reply:
Agreed, I'll add documentation for the possibility of running out of GIDs.

I couldn't think of a reliable way for now to recover the GIDs that have been used. I'll add a task to figure out and capture the used GIDs reliably in the event the driver is updated or restarted.